### PR TITLE
8341020: Error handler crashes when Metaspace is not fully initialized

### PR DESCRIPTION
**Pull request by shipilev**
When debugging CDS, I asked for `os::print_location` when Metaspace was not completely initialized. I suspect the same might happen if error handler fires very early in startup sequence.

`os::print_location` got to `Metaspace::contains`, which ended up in `Metaspace::is_in_nonclass_metaspace`, which crashed, since `VirtualSpaceList` was null. Other checks in `Metaspace::contains` seem to only check the address against the boundaries, and are not affected.

Additional testing:
 - [x] The same debugging session did not crash the VM after the patch

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341020](https://bugs.openjdk.org/browse/JDK-8341020): Error handler crashes when Metaspace is not fully initialized (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [7efa13b9](https://git.openjdk.org/jdk/pull/21206/files/7efa13b900639dc5e8e2d20db88af15905e798b9)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21206/head:pull/21206` \
`$ git checkout pull/21206`

Update a local copy of the PR: \
`$ git checkout pull/21206` \
`$ git pull https://git.openjdk.org/jdk.git pull/21206/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21206`

View PR using the GUI difftool: \
`$ git pr show -t 21206`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21206.diff">https://git.openjdk.org/jdk/pull/21206.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21206#issuecomment-2376722409)

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
LGTM!

Thanks

File: src/hotspot/share/memory/metaspace.cpp [hunk 1]
```diff
@@ -992,6 +992,7 @@ bool Metaspace::is_in_shared_metaspace(const void* ptr) {
 
 // Returns true if pointer points into one of the non-class-space metaspace regions.
 bool Metaspace::is_in_nonclass_metaspace(const void* ptr) {
-  return VirtualSpaceList::vslist_nonclass()->contains((MetaWord*)ptr);
+  VirtualSpaceList* list = VirtualSpaceList::vslist_nonclass();
+  return (list != nullptr) && list->contains((MetaWord*) ptr);
```
Comment (kimbarrett) [inline src/hotspot/share/memory/metaspace.cpp:996]: I'm not sure about this. This, and a number of other related functions nearby,
all seem like they should have Metaspace::initialized() as a precondition. And
the call to Metaspace::contains in os::print_location should be conditional on
Metaspace::initialized().  (There may be other "is initialized" checks that
should be made by (and offered to) os::print_location, but that's beyond the
scope of this change.)

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/memory/metaspace.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/memory/metaspace.cpp:996]: FWIW, we had a similar discussion in #19435.

Comment (jdksjolen) [reply src/hotspot/share/memory/metaspace.cpp:996]: All of these methods are public. I think that it seems sane to assume that they are the boundary for checking whether Metaspace is initialized before doing something.

Comment (shipilev) [reply src/hotspot/share/memory/metaspace.cpp:996]: All right, I think checking for `Metaspace::initialized()` in `os::print_location` is more in line with other changes, notably [JDK-8333290](https://bugs.openjdk.org/browse/JDK-8333290). This is Kim's suggestion, I think. So I did just that.

**Review by jdksjolen — APPROVED** [review]
Thanks, this looks good to me.

**Review by stefank — APPROVED** [review]
Looks good to me.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by shipilev** [comment]
Thank you!

/integrate

**Pull request closed without merging**